Bundler support for DRACOLoader and KTX2Loader (#26403)

cgauld opened the pull request:
Use the `new URL(..., import.meta.url);` structure, supported by browsers and several bundlers (including Webpack and Parcel), to allow for 'config-less' use of DRACOLoader and KTX2Loader

Related issue: #26403

cgauld committed:
Bundler support for DRACOLoader and KTX2Loader (#26403)

Use the `new URL(..., import.meta.url);` structure, supported by browsers and several bundlers (including Webpack and Parcel), to allow for 'config-less' use of DRACOLoader and KTX2Loader

gkjohnson commented:
I'll leave detailed feedback to others but in the mean time can we update one of the examples to remove the call to DracoLoader's `setDecoderPath` to easily demonstrate that this works with three.js vanilla js approach.

Also it would be best to use the code style consistent with the rest of the project. You can run `npm run lint-addons` to see the style issues in the files.

Thanks for making this change!

cgauld committed:
Update DRACO and KTX2 examples to remove the use of `setDecoderPath` and `setTranscoderPath`

cgauld committed:
Lint using eslint `ecmaVersion` of 2020 in order to correctly handle `import.meta.url`

cgauld committed:
Fix linting issues for DRACOLoader and KTX2Loader

cgauld commented:
Thanks for your fast feedback :-) I've added a commit that removes the `setDecoderPath` from the `loader/draco` example and the `setTranscoderPath` from the `loader/texture/ktx2` and verified that they still work correctly (using `npm run dev`).

I ran the linting but the only output for the files I've changed was:
```
Parsing error: Unexpected token import
```

This is because the `eslint` option `ecmaVersion` is currently set to `2018`, which doesn't understand the `import.meta.url` structure. I've added additional commits that update that value to an `ecmaVersion` of `2020` and to address the subsequently visible linting issues in my changes.

Mugen87 commented on .eslintrc.json:
```diff
@@ -6,7 +6,7 @@
   },
   "parserOptions": {
-    "ecmaVersion": 2018,
+    "ecmaVersion": 2020,
```
We want to stick to ECMA 2018 so please don't change this line. Add an exception via a comment to the lines where eslint complaints.

cgauld replied:
Unfortunately it's an eslint parsing error rather than a rule violation so the usual way to add an exception for the line doesn't work. The only alternative I know of would be to ignore these files as a whole in the eslint config, like this:
```json
  "ignorePatterns": [
    "examples/jsm/loaders/DRACOLoader.js",
    "examples/jsm/loaders/KTX2Loader.js"
  ],
```
But that does mean we don't get the benefit of linting for the entirety of those files. Let me know if you'd like me to proceed on that basis.

Mugen87 replied:
Then let's exclude both files from linting. 

BTW: What ECMA 2020 syntax are you using? We have to be aware that this syntax does not break bundlers like WebPack 4. That would be a red flag that makes the PR impossible to merge.

gkjohnson commented:
Live links to the changed examples:

- https://raw.githack.com/cgauld/three.js/bundler-support/examples/webgl_loader_draco.html
- https://raw.githack.com/cgauld/three.js/bundler-support/examples/webgl_loader_texture_ktx2.html

cgauld commented:
@Mugen87 make a good point regarding support for webpack 4. While webpack 5 does support `import.meta.url`, it looks like webpack 4 does not 😢. If support for that version is important for the project, than perhaps we should consider again the other approach where we have an optional `resolveDependencies` function that `DRACOLoader` and `KTX2Loader` use (if set). That would move any actual use of `import.meta.url` to user code, should they wish to use it. The disadvantage of course is that we lose significant simplicity afforded by this PR, where the user doesn't have to know about or handle the dependencies at all. Let me know your thoughts @gkjohnson!

gkjohnson commented:
> make a good point regarding support for webpack 4. While webpack 5 does support import.meta.url, it looks like webpack 4 does not

I forgot we were hoping to continue supporting Webpack 4. @Mugen87 how long should we expect to continue supporting the older version of Webpack? Webpack 5.0 is nearly 3 years old at this point and `import.meta` has [been supported in modern browsers since 2018](https://caniuse.com/?search=import.meta). Parcel, vite, and other modern bundlers support it now, as well. If it were my decision we would accept that it breaks some now-legacy build processes especially since these are example files that can be copied if needed. And if we get a lot of complaints and people are unable to upgrade we can consider a different option without using an `import.meta` default.

Mugen87 commented:
> how long should we expect to continue supporting the older version of Webpack

I don't know that. But when we discussed this topic half a year ago (see #24237), it was clearly too early to stop the support.

cgauld commented:
Thanks for your thoughts! I'll pull together a PR with the alternative so we can compare. We can always revisit the approach in this PR in the future once Webpack 4 support can be discontinued.

Makio64 commented:
> > how long should we expect to continue supporting the older version of Webpack
> 
> I don't know that. But when we discussed this topic half a year ago (see #24237), it was clearly too early to stop the support.
> 
> TBH, if we want to merge this PR in the foreseeable future, we need a ECMA 2018 compatible solution. I don't want to see complaints from the user side about this Webpack topic again.

Push people to more modern workflow is not that bad, switching from webpack v4 to v5 or Vite should take a day maximum and will make the devs/teams happier with better perf.

Maybe mark Webpack 4 and old bundlers as deprecated in the doc / readme will give the time for the teams to switch before a real drop of their support ? ( like its done with some features in three )

LeviPesin commented:
It seems from the [statistics](https://www.npmjs.com/package/webpack?activeTab=versions) that in the last week Webpack 5 received 16.7m downloads, while Webpack 4 received 8.5m (Webpack 0-3 combined received less than 0.5m). So I'm afraid it's still too early to drop the support for Webpack 4...

Mugen87 commented:
> We can always revisit the approach in this PR in the future once Webpack 4 support can be discontinued.

Okay, closing until this is the case.

CodyJasonBennett commented:
For the future, and this may be worth a separate issue, but I'd note that top-level await as used in WebGPU code removes the possibility of supporting Webpack 4 and CJS. It's also a bit more modern than ES2020 and `import.meta.url` so Vite will complain for the moment (it uses a browserslist for `optimizeDeps`). That might resolve itself in time if Apple drags their feet.
